aspnetcore/fundamentals/servers/kestrel.md

```diff
@@ -781,27 +751,82 @@ public class TlsFilterConnectionHandler : ConnectionHandler
         // requested algorithm(s).
         if (tlsFeature.CipherAlgorithm == CipherAlgorithmType.Null)
         {
-            throw new NotSupportedException("Prohibited cipher: " + tlsFeature.CipherAlgorithm);
+            throw new NotSupportedException("Prohibited cipher: " + 
+                tlsFeature.CipherAlgorithm);
         }
 
-        while (true)
+        var result = await connection.Transport.Input.ReadAsync();
```

Comment:
@halter73 this copy loop isn't needed, is it?

Reply:
I can see that I was too literal here. I see you want the *whole loop* dropped now.

I'm tired ... battling ⚔️ with integration testing bits for hours today. 😵

Reply:
I'm not sure why this sample is doing any I/O at all. It should call a `next` ConnectionDelegate as long as a non-null cypher algorithm is selected.

If there isn't going to be a `next` ConnectionDelegate, it probably makes sense to continue doing an echo server example by keeping the while loop. Unfortunately UseConnectionHandler doesn't support `next` atm.  @davidfowl

Reply:
This is the current pattern for using next w/ a ConnectionHandler-like type: 

```c#
            listenOptions.Use(next =>
            {
                // Set the list of protocols from listen options
                httpsOptions.HttpProtocols = listenOptions.Protocols;
                var middleware = new HttpsConnectionMiddleware(next, httpsOptions, loggerFactory);
                return middleware.OnConnectionAsync;
            });
```

https://github.com/aspnet/AspNetCore/blob/4a7bf75e89e3dfe7264a641ed041f50788d28d20/src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs#L220-L226

Reply:
Lets clean this up a bit and use a delegate directly. It'll be a bit cleaner.

```C#
using System;
using System.Security.Authentication;
using Microsoft.AspNetCore.Connections.Features;

namespace Microsoft.AspNetCore.Connections
{
    public static class TlsFilterConnectionMiddlewareExtensions
    {
        public static IConnectionBuilder UseTlsFilter(this IConnectionBuilder builder)
        {
            return builder.Use((connection, next) =>
            {
                var tlsFeature = connection.Features.Get<ITlsHandshakeFeature>();

                // Throw NotSupportedException for any cipher algorithm that the app doesn't
                // wish to support. Alternatively, define and compare
                // ITlsHandshakeFeature.CipherAlgorithm to a list of acceptable cipher
                // suites.
                //
                // A ITlsHandshakeFeature.CipherAlgorithm of CipherAlgorithmType.Null
                // indicates that no cipher algorithm supported by Kestrel matches the
                // requested algorithm(s).
                if (tlsFeature.CipherAlgorithm == CipherAlgorithmType.Null)
                {
                    throw new NotSupportedException("Prohibited cipher: " +
                        tlsFeature.CipherAlgorithm);
                }

                // Call the next connection middleware in the pipeline
                return next();
            });
        }
    }
}
```

Usage looks like:

```C#
webBuilder.ConfigureKestrel(serverOptions =>
{
    serverOptions.Listen(IPAddress.Any, 8000, listenOptions =>
    {
        listenOptions.UseHttps("testCert.pfx", "testPassword");
        listenOptions.UseTlsFilter();
    });
});
```